build/debs/build.go

```diff
@@ -241,177 +234,214 @@ func fetchVersion(url string) (string, error) {
+
+	return getStableKubeVersion()
+}
+
 func getStableKubeVersion() (string, error) {
 	return fetchVersion("https://dl.k8s.io/release/stable.txt")
 }
 
-func getLatestKubeVersion() (string, error) {
+func getTestingKubeVersion() (string, error) {
 	return fetchVersion("https://dl.k8s.io/release/latest.txt")
 }
 
-func getLatestCIVersion() (string, error) {
-	latestVersion, err := getLatestKubeCIBuild()
+func getUnstableKubeVersion() (string, error) {
+	latestCIVersion, err := getLatestKubeCIBuild()
 	if err != nil {
 		return "", err
 	}
 
-	// Replace the "+" with a "-" to make it semver-compliant
-	return strings.Replace(latestVersion, "+", "-", 1), nil
-}
-
-func getCRIToolsLatestVersion() (string, error) {
-	return criToolsVersion, nil
+	return latestCIVersion, nil
 }
 
 func getLatestKubeCIBuild() (string, error) {
 	return fetchVersion("https://dl.k8s.io/ci/k8s-master.txt")
 }
 
-func getCIBuildsDownloadLinkBase(_ version) (string, error) {
-	latestCiVersion, err := getLatestKubeCIBuild()
-	if err != nil {
-		return "", err
+func getDownloadLinkBase(packageDef packageDefinition) (string, error) {
+	switch packageDef.Distribution {
+	case DistributionUnstable:
+		return getCIBuildsDownloadLinkBase(packageDef)
 	}
 
-	return fmt.Sprintf("https://dl.k8s.io/ci/v%s", latestCiVersion), nil
+	return getReleaseDownloadLinkBase(packageDef)
 }
 
-func getReleaseDownloadLinkBase(v version) (string, error) {
-	return fmt.Sprintf("%s/v%s", releaseDownloadLinkBase, v.Version), nil
+func getCIBuildsDownloadLinkBase(packageDef packageDefinition) (string, error) {
+	ciVersion := packageDef.KubernetesVersion
+	if ciVersion == "" {
+		var err error
+		ciVersion, err = getLatestKubeCIBuild()
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return fmt.Sprintf("https://dl.k8s.io/ci/v%s", ciVersion), nil
 }
 
-func getKubeadmDependencies(v version) (string, error) {
-	cniVersion, err := getKubeletCNIVersion(v)
-	if err != nil {
-		return "", err
-	}
+func getReleaseDownloadLinkBase(packageDef packageDefinition) (string, error) {
+	return fmt.Sprintf("%s/v%s", releaseDownloadLinkBase, packageDef.KubernetesVersion), nil
+}
 
+func getKubeadmDependencies(packageDef packageDefinition) (string, error) {
 	deps := []string{
-		"kubelet (>= 1.13.0)",
-		"kubectl (>= 1.13.0)",
-		fmt.Sprintf("kubernetes-cni (%s)", cniVersion),
+		fmt.Sprintf("kubelet (>= %s)", minimumKubernetesVersion),
+		fmt.Sprintf("kubectl (>= %s)", minimumKubernetesVersion),
+		fmt.Sprintf("kubernetes-cni (>= %s)", minimumCNIVersion),
+		fmt.Sprintf("cri-tools (>= %s)", minimumCRIToolsVersion),
 		"${misc:Depends}",
 	}
-	sv, err := semver.Make(v.Version)
+
+	return strings.Join(deps, ", "), nil
+}
+
+func getCNIVersion(packageDef packageDefinition) (string, error) {
+	log.Printf("using CNI version")
+
+	kubeSemver, err := semver.Make(packageDef.KubernetesVersion)
 	if err != nil {
 		return "", err
 	}
 
-	v1110, err := semver.Make("1.11.0-alpha.0")
+	v117, err := semver.Make("1.17.0-alpha.0")
```

Comment:
Why is this configuration hard coded as a string constant and further encoded in variable naming?